src/linux_ip_socket/linux_ip_socket.cpp

```diff
@@ -0,0 +1,336 @@
+    if(sockfd == -1) {
+        std::cerr << "socket() returned an error: " << strerror(errno) << std::endl;
+        freeaddrinfo(address_array);
+        return INVALID_SOCKET_ID;
+    }
+    const int connect_result = connect(sockfd, connect_address->ai_addr, connect_address->ai_addrlen);
+    if(connect_result == -1) {
+        std::cerr << "connect() returned an error: " << strerror(errno) << std::endl;
+        freeaddrinfo(address_array);
+        return INVALID_SOCKET_ID;
+    }
+
+    freeaddrinfo(address_array);
+    return sockfd;
+}
+
+void
+linux_ip_socket_private_data::prepare_listen_socket()
+{
+    addrinfo* address_array = nullptr;
+    find_addresses(&address_array, m_ip_device_configuration->address, m_ip_device_configuration->port);
+
+    addrinfo* listen_address = nullptr;
+    for(listen_address = address_array; listen_address != nullptr; listen_address = listen_address->ai_next) {
+        m_listen_sockfd = socket(listen_address->ai_family, listen_address->ai_socktype, listen_address->ai_protocol);
+        if(m_listen_sockfd == -1) {
+            std::cerr << "socket() returned an error: " << strerror(errno) << std::endl;
+            continue;
+        }
+        const int bind_result = bind(m_listen_sockfd, listen_address->ai_addr, listen_address->ai_addrlen);
+        if(bind_result == -1) {
+            std::cerr << "bind() returned an error: " << strerror(errno) << std::endl;
+            continue;
+        }
+
+        break;
+    }
+
+    if(listen_address == nullptr) {
+        std::cerr << "Cannot create or bind socket\n";
+        abort();
+    }
+
+    freeaddrinfo(address_array);
+
+    const int listen_result = listen(m_listen_sockfd, DRIVER_MAX_CONNECTIONS);
+    if(listen_result == -1) {
+        std::cerr << "Cannot listen on socket\n";
+        abort();
+    }
+}
+
+void
+linux_ip_socket_private_data::initialize_packet_parser()
+{
+    m_parse_state = STATE_WAIT;
+}
+
+void
+linux_ip_socket_private_data::accept_connection(pollfd* table, int& table_size)
+{
+    sockaddr_storage remote_addr;
+    socklen_t remote_addr_size = sizeof(sockaddr_storage);
+    const int new_sockfd = accept(m_listen_sockfd, (struct sockaddr*)&remote_addr, &remote_addr_size);
+    if(new_sockfd == -1) {
+        std::cerr << "accept() returned an error: " << std::strerror(errno) << std::endl;
+    } else {
+        initialize_packet_parser();
+        table[table_size].fd = new_sockfd;
+        table[table_size].events = POLLIN;
+        ++table_size;
+    }
+}
+
+void
+linux_ip_socket_private_data::read_data_or_disconnect(const int table_index, pollfd* table, int& table_size)
+{
+    const int recv_result = recv(table[table_index].fd, m_recv_buffer, DRIVER_RECV_BUFFER_SIZE, 0);
+    if(recv_result == -1) {
+        std::cerr << "recv() returned an error: " << std::strerror(errno) << std::endl;
+    } else if(recv_result == 0) {
+        --table_size;
+        close(table[table_size].fd);
```

Comment:
Are we guaranteed to connect and disconnect in reverse order? We receive on table_index, but close on table_size - 1....

Reply:
So, in this moment only one connection makes sense. Multiple connections require multiple buffers - at this moment only one is available.

Reply:
OK... so is closing table[table_size].fd valid after receiving from table[table_index].fd?

Reply:
I'm not sure if I understand the question.
The `recv` (https://linux.die.net/man/3/recv) returns zero if the other side closes the connection - so, for cleanup `close(table_size].fd)` shall be called. Ok, I have an idea to refactor this, since the only one connection is allowed, then no table is required.

Reply:
Code was refactored